### PR TITLE
Add possibility to set separate xml coverage and junit log

### PR DESCRIPTION
**Pull request by vkhramtsov**
This PR:

- [x] Adds new feature `Separate options for path to xml coverage and junit log`
- [ ] Covered by tests
- [ ] Doc PR: https://github.com/infection/site/pull/XXX

Fixes https://github.com/infection/infection/issues/1377

### CODE REVIEW

**Comment by vkhramtsov** [comment]
@maks-rafalko @sanmai Is it ok to implement functionality like this? I would like to have feedback before implementing tests and adjusting documentation.

File: src/Command/RunCommand.php
Comment (sanmai) [inline src/Command/RunCommand.php:77]: These could follow the style of PHPUnit: `--coverage-xml` `--log-junit`

What do you think?

File: src/Command/RunCommand.php
Comment (sanmai) [inline src/Command/RunCommand.php:162]: I don't mind constants, and I'd probably use them in a similar situation, but this goes out of line with other parts of the file tbh.

But don't take just my word for it. Let's hear what others think.

File: src/Container.php
Comment (sanmai) [inline src/Container.php:272]:
```php
$configuration = $container->getConfiguration();

$configuration->getXmlCoveragePath() ?? $configuration->getCoveragePath();
```
?

**Comment by sanmai** [comment]
This definitely needs an E2E test. It should be not very hard to add with `tests/add_new_e2e` FWIW.

File: src/Command/RunCommand.php
Comment (sanmai) [inline src/Command/RunCommand.php:414]: ?

File: src/Container.php
Comment (sanmai) [inline src/Container.php:767]: Is this used anywhere?

File: src/Command/RunCommand.php
Comment (vkhramtsov) [reply src/Command/RunCommand.php:77]: I'm implementing this according to description in ticket https://github.com/infection/infection/issues/1377. I could change this without any problem if you need. Your project, your rules :wink:

Comment (sanmai) [reply src/Command/RunCommand.php:77]: What do you think will be most sensible, to use some original keys, or use the same keys PHPUnit uses to generate the very reports?

File: src/Command/RunCommand.php
Comment (vkhramtsov) [reply src/Command/RunCommand.php:162]: I can refactor all command. From my education and previous experience: in case of using the same value more than one in same context -- it should be variable or constant. This could help with decreasing number of mistakes. But anyway, as usual: your project -- your rules :wink:

File: src/Command/RunCommand.php
Comment (vkhramtsov) [reply src/Command/RunCommand.php:414]: My mistake, I'm going to remove this

File: src/Command/RunCommand.php
Comment (sanmai) [reply src/Command/RunCommand.php:162]: I agree, but a hodgepodge of constants and string literals is, well, confusing. One starts asking why there's a constant for this, and not a constant for that... In this situation it might be more important to avoid confiscation over following the general (correct) rule for constants. But let's hear out other reviewers.

I don't think it'll be a terrible idea to convert this file to use constants for all commands, but surely in a different, much shorter PR.

File: src/Command/RunCommand.php
Comment (vkhramtsov) [reply src/Command/RunCommand.php:77]: It makes sense to use the same parameter names as in phpunit. I'm going to change this.

File: src/Command/RunCommand.php
Comment (vkhramtsov) [reply src/Command/RunCommand.php:162]: Good point for second PR :wink:

File: src/Command/RunCommand.php
Comment (sanmai) [reply src/Command/RunCommand.php:77]: That's a good point about the project, but that's _your_ pull request. Therefore:
- It's _your_ choice how it's going to be. If you make it good, the entire project will be better. E.g. assuming you're using Infection, if you'd make it a poo, you'll be using a poo. Let's focus on not making it worse, but better, together.
- I'm not an Einstein-like genius, I can't know the true reasons you want to use this or that name here, so if you can say there's a reason to do this that way or another, then it could be just so. I will stand corrected then.

**Comment by vkhramtsov** [comment]
> This definitely needs an E2E test. It should be not very

I prefer to do this after getting full feedback. Also I need to add phpunit tests. Otherwise I have to change not only code, but tests also

File: src/Container.php
Comment (vkhramtsov) [reply src/Container.php:767]: [Here](https://github.com/vkhramtsov/infection/blob/feature-adjust-handling-of-predefined-xml-coverage-and-junit-log/src/Container.php#L793)?

Comment (sanmai) [reply src/Container.php:767]: What do you think about doing this in `getDefaultJUnitFilePath`?

File: src/Container.php
Comment (vkhramtsov) [reply src/Container.php]: Changed

File: src/Command/RunCommand.php
Comment (vkhramtsov) [reply src/Command/RunCommand.php]: @sanmai  I will create new issue for this.

File: src/Container.php
Comment (vkhramtsov) [reply src/Container.php:763]: From my point of view it will be overhead to get configuration from container in `getDefaultJUnitFilePath` method, but if you insisting...

Comment (sanmai) [reply src/Container.php:763]: Is it not memoized there is it?

File: src/Configuration/Schema/SchemaConfigurationFactory.php
Comment (sanmai) [inline src/Configuration/Schema/SchemaConfigurationFactory.php:71]: Variable names were on point tbh.

Comment (vkhramtsov) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: @sanmai Sorry, I don't understand you. Would you be so kind to add more context here?

Comment (sanmai) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: The names like `xmlCoveragePath` you were using were excellent and on point.

File: src/Container.php
Comment (vkhramtsov) [reply src/Container.php:763]: Let's wait for review from @maks-rafalko. After that I will change this in case it will be still actual.

File: src/Configuration/Schema/SchemaConfigurationFactory.php
Comment (vkhramtsov) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: Changing variable here means that I need to change name of setting in configuration file, but I prefer to have the same parameters in CLI and configuration, so I prefer to leave it as is here.

Comment (sanmai) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: I can't agree with you here. CLI parameters are such because of PHPUnit. Here we're not limited by the same constraint.

Comment (vkhramtsov) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: Waiting for review from @maks-rafalko

**Comment by vkhramtsov** [comment]
PR rebased against master. I have several warnings about non covered functionality and several issues in non changed code.

File: src/Configuration/Schema/SchemaConfigurationFactory.php
Comment (sanmai) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: I'd like to give this a green light, but this part is what I really like to see fixed.

Comment (vkhramtsov) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: We need third opinion. I really don't like situation, when we have CLI options and configuration settings differently named. Problem it that you need to read more documentation and it could be additional point of failure, because you cannot easily transfer data from command line to config, because you have to alter it

Comment (sanmai) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: What you did here is really great, and I'd say with this small naming issue fixed it'll all will be just _perfect_.

Comment (vkhramtsov) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: @sanmai another option is revert command line parameters name to initial in this PR.

**Comment by vkhramtsov** [comment]
Should I do something with automatic comments, related to not changed code?

File: src/Configuration/Schema/SchemaConfigurationFactory.php
Comment (sanmai) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: This is a truly despicable option tbh.

**Comment by sanmai** [comment]
Notices for unchanged code usually not a cause of concerns, especially if tests weren't changed too.

File: src/Configuration/Schema/SchemaConfigurationFactory.php
Comment (vkhramtsov) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: @sanmai so, are you insisting on changing configuration file settings? In this case I'm going to rollback CLI options to their initial names because I want to have consistency between configuration and command line.

**Review by sanmai — CHANGES_REQUESTED** [review]
I see why it is good to use PHPUnit-style CLI options, but I see no good reason to use `$rawConfig->xmlCoverage` and `$rawConfig->logJunit` instead of easy-to-read and clear-to-understand  `$rawConfig->xmlCoveragePath` and `$rawConfig->junitLogPath`.

File: src/Configuration/Schema/SchemaConfigurationFactory.php
Comment (sanmai) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: I'd rather see both of these changes happen simultaneously. I don't understand what we're arguing about here. You can check other pull requests to see that purposeless nitpicking isn't something I find joy in. Neither I fancy spending the little free time I have arguing without arguments. ("I really don't like situation" isn't sound and neither an argument, but an appeal to a subjective taste in things.)

Therefore, this will be my final review of this PR unless changes are made, or there a _sound_ argument. Thank you.

Comment (maks-rafalko) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: Difficult question ;) Let me share some thoughts/facts on this topic:

1. From what I see, we don't have any example when CLI option is named differently that corresponded config
2. we have many inconsistencies in **already existing** config settings and proposed settings in this PR (`logJunit` or `junitLogPath`

here is what I mean:

* `bootstrap` config settings - it's a path to a file, but it's not `bootstrapPath` (bad)
* `tmpDir` / `configDir` - this is a path to **directory**, hence `***Dir` postfix

What is proposed here by @sanmai:

* `xmlCoveragePath` for a **dir**
* `junitLogPath` for a **file**

As you can see, neither existing config in `master` branch, nor proposed names are correc. They are not consistent (`bootstrap` has no `Path` postfix, `xmlCoveragePath` should be `xmlCoverageDir`, and third issue - CLI options are named differently

What I can suggest: we must agree on the following:

1. if it's a path - postfix must be `***Path`
2. if it's a dir - postfix must be `***Dir`
3. what to do with different names of CLI options and config settings

Personally, 

* I find `xmlCoverageDir` and `junitLogPath` much more self-explaining thatn variants with `Path`/`Dir` postfix
* I find naming CLI options and config settings in a similar way - more clear than naming them differently. But if you both really like PHPUnit-like names (`coverage-xml`, `log-junit`) - let it be so.

I didn't get why don't you like the following approach:

* `--xml-coverage-dir` and `xmlCoverageDir` config setting
* `--junit-log-path` and `junitLogPath` config setting

One more thing to note: Infection is not only about PHPUnit, it's about PHPSpec and Codeception as well, which simply don't have `--coverage-xml` and `--log-junit` settings

File: src/Container.php
Comment (maks-rafalko) [reply src/Container.php:780]: Here I agree with @sanmai - it's better to move it to `getDefaultJUnitFilePath()`. The logic will be in one place (method) and we won't have state mutation outside of `getDefaultJUnitFilePath()` - it will be just memoization saved into the object state.

**Review by maks-rafalko — COMMENTED** [review]
Hello, sorry for the delay.

First of all, thank you for creating this PR and working on this feature!

Overall, looks good. Left some comments for the arguable places.

File: src/Configuration/Schema/SchemaConfigurationFactory.php
Comment (sanmai) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: Let me give it a shot an explaining.  Take this textbook example on using Infection with PHPUnit at CI:

```
vendor/bin/phpunit --coverage-xml=build/logs/coverage-xml --log-junit=build/logs/junit.xml
vendor/bin/infection --coverage=build/logs ... (skipping other options for brevity)
```
Now, let's imagine we're using these new options, but named not as in PHPUnit:
```
vendor/bin/phpunit --coverage-xml=build/logs/coverage-xml --log-junit=build/logs/junit.xml
vendor/bin/infection --xml-coverage-dir=build/logs/coverage-xml --junit-log-path=build/logs/junit.xml
```
Well, you can kinda see what's going where and why. How likely you will remember these? I don't think you will, probably will need to Google up. Which means these two new options not very far away from the old single option, usability-wise.

But what if we'll be using these new options named as they are in PHPUnit:
```
vendor/bin/phpunit --coverage-xml=build/logs/coverage-xml --log-junit=build/logs/junit.xml
vendor/bin/infection --coverage-xml=build/logs/coverage-xml --log-junit=build/logs/junit.xml 
```
Boom! You already remember how to use them. You will probably write them out of your mind. That's a different kind of UI. 

Make sense?

**Comment by sanmai** [comment]
No matter how I would like to have the options named like in PHPUnit, having the properties named consistently is far important for me. And neither pointless arguing back and forth is to my liking. Therefore, I propose to proceed as @maks-rafalko suggested, merge it, and move on. If'd find a gumption to propose [to alias this options to their PHPUnit counterparts](https://github.com/infection/infection/pull/1388#discussion_r516361650), then so to be, but if not I will have no one to blame.

File: src/Configuration/Schema/SchemaConfigurationFactory.php
Comment (maks-rafalko) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: makes sense

File: src/Command/RunCommand.php
Comment (maks-rafalko) [reply src/Command/RunCommand.php]: still not fixed

**Review by maks-rafalko — APPROVED** [review]
So, I'm :+1: with the current state of this PR, keeping in mind that we will make options/configuration settings consistent in a separate PR at some point later, as was discussed here https://github.com/infection/infection/pull/1388#discussion_r516251293

@vkhramtsov could you please proceed with Doc PR?

**Comment by vkhramtsov** [comment]
@maks-rafalko not so fast, wait a little bit. I need to find some time for testing and fixing open questions here. After that I'm going to add tests and documentation to this PR and pass review once again

File: src/Command/RunCommand.php
Comment (vkhramtsov) [reply src/Command/RunCommand.php]: Sure, should I remove phpdoc for all constants here?

File: src/Container.php
Comment (vkhramtsov) [reply src/Container.php]: Changed

File: src/Configuration/Schema/SchemaConfigurationFactory.php
Comment (vkhramtsov) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: So, what should I do here?

File: src/Command/RunCommand.php
Comment (vkhramtsov) [reply src/Command/RunCommand.php]: @maks-rafalko should I remove phpdoc for all constants in this file?

Comment (maks-rafalko) [reply src/Command/RunCommand.php]: yes please

Comment (vkhramtsov) [reply src/Command/RunCommand.php]: Removed

File: src/Configuration/Schema/SchemaConfigurationFactory.php
Comment (vkhramtsov) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: @maks-rafalko @sanmai What should I do with parameters and configuration file settings? Should I leave them in current state or change?
After finishing with this question, I'm going to proceed with documentation and tests.
Thank you.

Comment (maks-rafalko) [reply src/Configuration/Schema/SchemaConfigurationFactory.php:71]: as for me, let's proceed with the current state: `log-junit` and `coverage-xml` names, same as for PHPUnit.

Other inconsistencies described above should/will be addressed in another PR

**Comment by vkhramtsov** [comment]
Unfortunately I don't have time to work with this feature. I'm closing this PR.

**Pull request closed without merging**
